python/ray/serve/application.py

```diff
@@ -0,0 +1,319 @@
+
+        Raises:
+            ValueError: If a deployment with deployment.name already exists in
+                this Application.
+        """
+
+        if not isinstance(deployment, Deployment):
+            raise TypeError(f"Got {type(deployment)}. Expected deployment.")
+        elif deployment.name in self._deployments:
+            raise ValueError(
+                f'Deployment with name "{deployment.name}" already '
+                "exists in this application. To overwrite this "
+                "deployment, use attribute or index notation. "
+                "For example:\n\napp.deployment_name = "
+                "new_deployment"
+            )
+
+        self._deployments[deployment.name] = deployment
+
+    def deploy(self, blocking: bool = True):
+        """Atomically deploys the Application's deployments to the Ray cluster.
+
+        The Application's deployments can carry handles to one another.
+
+        Args:
+            blocking (bool): If True, this function only returns after the
+                deployment is finished. If False, this function returns
+                immediately after requesting the deployment.
+        """
+
+        if len(self._deployments) == 0:
+            return
+
+        parameter_group = []
+
+        for deployment in self._deployments.values():
+            if not isinstance(deployment, Deployment):
+                raise TypeError(
+                    f"deploy_group only accepts Deployments, but got unexpected "
+                    f"type {type(deployment)}."
+                )
+
+            deployment_parameters = {
+                "name": deployment._name,
+                "func_or_class": deployment._func_or_class,
+                "init_args": deployment.init_args,
+                "init_kwargs": deployment.init_kwargs,
+                "ray_actor_options": deployment._ray_actor_options,
+                "config": deployment._config,
+                "version": deployment._version,
+                "prev_version": deployment._prev_version,
+                "route_prefix": deployment.route_prefix,
+                "url": deployment.url,
+            }
+
+            parameter_group.append(deployment_parameters)
+
+        return _get_global_client().deploy_group(parameter_group, _blocking=blocking)
+
+    def run(self, logger=logger):
+        """Deploys all deployments in this Application and logs status.
+
+        This function keeps looping and printing status, so it must be manually
+        killed (e.g. using ctrl-C). When it recieves a kill signal, it tears
+        down all deployments that it deployed. If there are no deployments
+        left, it also shuts down the rest of Serve, including the controller.
+        This is meant to help interactive development.
+
+        Args:
+            logger: Any Python object that implements the standard logger
+                interface.
+        """
+
+        try:
+            serve.start(detached=True)
```

Comment:
you can remove this line, the serve API calls will do this implicitly if it hasn't been called yet.

Reply:
I haven't dug deep, but the unit tests fail without `serve.start(detached=True)` there, so I'll keep it in.